File: src/main/java/tech/bobliu/assignment08/service/CaptchaServiceImpl.java
package tech.bobliu.assignment08.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Service;
import tech.bobliu.assignment08.config.CaptchaProperties;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class CaptchaServiceImpl implements CaptchaService {

    private final List<Captcha> captchaList;

    @Autowired
    public CaptchaServiceImpl(CaptchaProperties captchaProperties) {
        this.captchaList = captchaProperties.getList().stream()
                .map(it -> new Captcha(it.getFilename(), it.getResult()))
                .collect(Collectors.toList());
    }

    @Override
    public CaptchaResponse getNextCaptcha(Integer lastIndex) {
        int captchaIndex;
        do {
            captchaIndex = (int) (Math.random() * captchaList.size());
        } while (lastIndex != null && captchaIndex == lastIndex);

        return new CaptchaResponse(captchaList.get(captchaIndex), captchaIndex);
    }

    @Override
    public InputStream getCaptchaImage(String filename) {
        ClassPathResource resource = new ClassPathResource("captcha/" + filename);
        try {
            return resource.getInputStream();
        } catch (IOException e) {
            throw new RuntimeException("Failed to load captcha image", e);
        }
    }
}

File: src/main/kotlin/tech/bobliu/assignment08/service/CaptchaServiceImpl.kt
package tech.bobliu.assignment08.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Service
import tech.bobliu.assignment08.config.CaptchaProperties
import java.io.InputStream

@Service
class CaptchaServiceImpl @Autowired constructor(
    private val captchaProperties: CaptchaProperties
) : CaptchaService {

    private val captchaList = captchaProperties.list.map { 
        CaptchaService.Captcha(it.filename, it.result) 
    }

    override fun getNextCaptcha(lastIndex: Int?): Pair<CaptchaService.Captcha, Int> {
        var captchaIndex: Int
        do {
            captchaIndex = (Math.random() * captchaList.size).toInt()
        } while (lastIndex != null && captchaIndex == lastIndex)

        return Pair(captchaList[captchaIndex], captchaIndex)
    }

    override fun getCaptchaImage(filename: String): InputStream {
        val resource = ClassPathResource("captcha/$filename")
        return resource.inputStream
    }
}
